src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/SegmentedList`1.cs

```diff
@@ -0,0 +1,794 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using IEnumerable = System.Collections.IEnumerable;
+using IEnumerator = System.Collections.IEnumerator;
+
+namespace Microsoft.CodeAnalysis.Shared.Collections
+{
+    internal interface IByRefComparer<T>
+    {
+        int Compare(in T first, in T second);
+    }
+
+    /// <summary>
+    /// Segmented list implementation, copied from Microsoft.Exchange.Collections.
+    /// </summary>
+    /// <typeparam name="T">The type of the list element.</typeparam>
+    /// <remarks>
+    /// This class implement a list which is allocated in segments, to avoid large lists to go into LOH.
+    /// </remarks>
+    internal sealed class SegmentedList<T> : ICollection<T>, IReadOnlyList<T>
+    {
+        private readonly int _segmentSize;
+        private readonly int _segmentShift;
+        private readonly int _offsetMask;
+
+        private int _capacity;
+        private int _count;
+        private T[][] _items;
+
+        /// <summary>
+        /// Constructs SegmentedList.
+        /// </summary>
+        /// <param name="segmentSize">Segment size</param>
+        public SegmentedList(int segmentSize)
+            : this(segmentSize, 0)
+        {
+        }
+
+        /// <summary>
+        /// Constructs SegmentedList.
+        /// </summary>
+        /// <param name="segmentSize">Segment size</param>
+        /// <param name="initialCapacity">Initial capacity</param>
+        public SegmentedList(int segmentSize, int initialCapacity)
+        {
+            if (segmentSize <= 1 || (segmentSize & (segmentSize - 1)) != 0)
+            {
+                throw new ArgumentOutOfRangeException("segment size must be power of 2 greater than 1");
+            }
+
+            _segmentSize = segmentSize;
+            _offsetMask = segmentSize - 1;
+            _segmentShift = 0;
+
+            while (0 != (segmentSize >>= 1))
+            {
+                _segmentShift++;
+            }
+
+            if (initialCapacity > 0)
+            {
+                initialCapacity = _segmentSize * ((initialCapacity + _segmentSize - 1) / _segmentSize);
+                _items = new T[initialCapacity >> _segmentShift][];
+                for (var i = 0; i < _items.Length; i++)
+                {
+                    _items[i] = new T[_segmentSize];
+                }
+
+                _capacity = initialCapacity;
+            }
+        }
+
+        /// <summary>
+        /// Returns the count of elements in the list.
+        /// </summary>
+        public int Count
+        {
+            get { return _count; }
+        }
+
+        /// <summary>
+        /// Copy to Array
+        /// </summary>
+        /// <returns>Array copy</returns>
+        public T[] UnderlyingArray => ToArray();
```

Comment:
why do we need both UnderlyingArray and ToArray?  The name is also misleading.  Underlying makes it sounds like the actual internal array is being returned.